checker/src/org/checkerframework/checker/lock/LockVisitor.java

```diff
@@ -4,88 +4,366 @@
 import com.sun.source.tree.MethodTree;
+import com.sun.source.tree.SynchronizedTree;
 import com.sun.source.tree.Tree;
+import com.sun.source.tree.VariableTree;
+import com.sun.source.tree.Tree.Kind;
 import com.sun.source.util.TreePath;
 
 /**
- * The LockVisitor enforces the subtyping rules of LockHeld and LockPossiblyHeld
- * (via BaseTypeVisitor). It also manually verifies that @Holding and @HoldingOnEntry
- * annotations are properly used on overridden methods. It handles @GuardedBy annotations
- * on method receivers. Finally, it ensures that we avoid doing any lock checking
- * when visiting initializers.
+ * The LockVisitor enforces the special type-checking rules described in the Lock Checker manual chapter.
+ *
+ * @checker_framework.manual #lock-checker Lock Checker
  */
 
+// TODO: Enforce that lock expressions are final or effectively final.
+
 public class LockVisitor extends BaseTypeVisitor<LockAnnotatedTypeFactory> {
-    private final Class<? extends Annotation> checkerGuardedByClass = org.checkerframework.checker.lock.qual.GuardedBy.class;
-    private final Class<? extends Annotation> checkerHoldingClass = org.checkerframework.checker.lock.qual.Holding.class;
-    private final Class<? extends Annotation> checkerHoldingOnEntryClass = org.checkerframework.checker.lock.qual.HoldingOnEntry.class;
-    private final Class<? extends Annotation> checkerLockHeldClass = org.checkerframework.checker.lock.qual.LockHeld.class;
+    private final Class<? extends Annotation> checkerGuardedByClass = GuardedBy.class;
+    private final Class<? extends Annotation> checkerGuardSatisfiedClass = GuardSatisfied.class;
 
-    // Note that Javax and JCIP @GuardedBy is used on both methods and objects. For methods they are
-    // equivalent to the Checker Framework @Holding annotation.
-    private final Class<? extends Annotation> javaxGuardedByClass = javax.annotation.concurrent.GuardedBy.class;
-    private final Class<? extends Annotation> jcipGuardedByClass = net.jcip.annotations.GuardedBy.class;
+    /** Annotation constants */
+    protected final AnnotationMirror GUARDEDBY, GUARDEDBYUNKNOWN, GUARDSATISFIED, GUARDEDBYBOTTOM;
 
     public LockVisitor(BaseTypeChecker checker) {
         super(checker);
 
+        GUARDEDBYUNKNOWN = AnnotationUtils.fromClass(elements, GuardedByUnknown.class);
+        GUARDEDBY = AnnotationUtils.fromClass(elements, GuardedBy.class);
+        GUARDSATISFIED = AnnotationUtils.fromClass(elements, GuardSatisfied.class);
+        GUARDEDBYBOTTOM = AnnotationUtils.fromClass(elements, GuardedByBottom.class);
+
         checkForAnnotatedJdk();
     }
 
     @Override
+    public Void visitVariable(VariableTree node, Void p) { // visit a variable declaration
+        // A user may not annotate a primitive type, a boxed primitive type or a String
+        // with any qualifier from the @GuardedBy hierarchy.
+
+        // TODO: Note that there is currently no way to reliably retrieve user-written
```

Comment:
It does matter if the user wrote @GuardedBy({}). That is the default for primitives, but we forbid users from writing it explicitly because we will eventually introduce a @GuardedByName annotation appropriate for primitives. The warning that is issued here will be kept, but the message text can then change to "use @GuardedByName instead of @GuardedBy".
Should I keep the check here for the user-written @GuardedBy({}) and move the rest to BaseTypeValidator?